dataset/bucket.cpp

```diff
@@ -114,6 +116,58 @@ Dataset resize_default_init(const DatasetConstView &parent, const Dim dim,
   }
   return buffer;
 }
+
+template <class T>
+Variable from_constituents_impl(Variable &&indices, const Dim dim, T &&buffer) {
+  return {std::make_unique<variable::DataModel<bucket<T>>>(
+      std::move(indices), dim, std::move(buffer))};
+}
+
+Variable from_constituents(Variable indices, const Dim dim, DataArray buffer) {
+  return from_constituents_impl(std::move(indices), dim, std::move(buffer));
+}
+
+Variable from_constituents(Variable indices, const Dim dim, Dataset buffer) {
+  return from_constituents_impl(std::move(indices), dim, std::move(buffer));
+}
+
+namespace {
+template <class T> Variable bucket_sizes_impl(const VariableConstView &view) {
+  const auto &indices = std::get<0>(view.constituents<bucket<T>>());
+  const auto [begin, end] = unzip(indices);
+  return end - begin;
+}
+} // namespace
+
+Variable bucket_sizes(const VariableConstView &var) {
+  if (var.dtype() == dtype<bucket<Variable>>)
+    return bucket_sizes_impl<Variable>(var);
+  else if (var.dtype() == dtype<bucket<DataArray>>)
+    return bucket_sizes_impl<DataArray>(var);
+  else if (var.dtype() == dtype<bucket<Dataset>>)
+    return bucket_sizes_impl<Dataset>(var);
+  else
+    return makeVariable<scipp::index>(var.dims());
+}
+
+DataArray bucket_sizes(const DataArrayConstView &array) {
+  return {bucket_sizes(array.data()), array.aligned_coords(), array.masks(),
```

Comment:
I'm struggling to understand here why we pass back the unmodified coords and masks along with the bucket sizes. In this case are the aligned coords even guaranteed to be the same length as the new bucket sizes variable?

Reply:
Yes, think of this just like any other element-wise unary operation, e.g., `sin()`. The operation is simply applied to each data element, and a new data arrays is returned. In the case of `sin` we transform `array<double> -> array<double>`. `bucket_sizes` transforms `array<DataArrayView> -> array<double>`.

Reply:
Thanks that makes sense now.